### PR TITLE
Refactor Softmax

### PR DESCRIPTION
**Pull request by daemyung**
## 🙏 Describe the pull request

Softmax를 리펙토링 했습니다.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: trident/operation/softmax.py [hunk 1]
```diff
@@ -25,68 +26,71 @@ def forward(ctx: Any, *args: Any, **kwargs: Any):
         input, dim = args
         output = Softmax.__forward(input, dim)
 
-        ctx.save_for_backward(output)
+        ctx.save_for_backward(input, output)
```
Comment (kakao-steve-ai) [inline trident/operation/softmax.py]: input을 메모리에 저장하는것은 비효율적으로 보입니다. 필요한것은 dtype정도인거 같은데, 이것만 저장하는게 좋아보입니다.

Comment (daemyung) [reply trident/operation/softmax.py]: 비효율적이지 않습니다. 여기서는 레퍼런스만 들고있고 실제 메모리는 1개입니다. `grad_input`을 생성해주려면 `input`을 사용해서 생성하는게 가독성에 좋습니다.

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: `save_for_backward`를 하면 전체 텐서값을 저장하는거 아닌가요?

Comment (daemyung) [reply trident/operation/softmax.py]: 텐서의 레퍼런스를 저장하지 깊은 복사가 발생하지는 않습니다. 만약 깊은 복사가 매번 일어나면 메모리 터져요.

Comment (kakao-steve-ai) [reply trident/operation/softmax.py]: 레퍼런스만 저장해도 레퍼런스 카운트가 올라가서 메모리가 free 안되기 때문에 비효율적일거 같습니다. 메모리 free 타이밍을 정확히 모르겠지만, `backward`때까지 계속 들고 있을 필요가 없는 메모리를 들고 있어야 하니까요.

File: benchmarks/benchmark_softmax.py [hunk 2]
```diff
@@ -19,43 +19,31 @@
...
-    ["vec_sz"],
-    [256 * i for i in range(1, 21)],
-    {"num_bt": 32},
-)
-def bench_softmax_backward(num_bt, vec_sz, backend):
-    inp = torch.randn(num_bt, vec_sz, device="cuda", requires_grad=True)
+@util.report("softmax backward", ["x_size"], [2048 * i for i in range(1, 11)], {"y_size": 16})
+def bench_softmax_backward(y_size, x_size, backend):
+    input = torch.randn(y_size, x_size, device="cuda", requires_grad=True)
+    grad_output = torch.ones_like(input)
```
Comment (mejai1206) [inline benchmarks/benchmark_softmax.py]: `grad_output` 생성시 `randn` 으로 생성할 필요 없을까요?

File: trident/operation/softmax.py (hunk 1 above)
Comment (daemyung) [reply trident/operation/softmax.py]: 맞는말인데 훈련시 입력은 메모리 해제가 되지 않습니다.

File: benchmarks/benchmark_softmax.py (hunk 2 above)
Comment (daemyung) [reply benchmarks/benchmark_softmax.py]: 맞습니다. 수정하겠습니다.

File: trident/operation/attention.py [hunk 3]
```diff
@@ -132,13 +132,7 @@ def __backward(
...
         num_batches, num_heads, y_size, x_size = output.shape
         kernel.Softmax.backward_delta[(num_batches * num_heads * y_size,)](
-            delta,
-            output,
-            grad_output,
-            x_size=x_size,
-            y_size=num_batches * num_heads * y_size,
-            x_stride=1,
-            y_stride=x_size,
+            delta, output, grad_output, num_batches * num_heads * y_size, x_size, x_size, 1, util.dtype(delta.dtype)
```
Comment (mejai1206) [inline trident/operation/attention.py:135]: `x_size`가 커널 내부 `x_size`, `y_stride` 로써 두번 입력이 되는데, 하나만 보내는 방식은 좋지 않으려나요?

File: trident/kernel/softmax.py [hunk 4]
```diff
@@ -20,226 +20,183 @@
...
-            grad_output = tl.load(grad_output_block_ptr, boundary_check=(1,), padding_option="zero").to(tl.float32)
-            delta += output * grad_output
-            output_block_ptr = tl.advance(output_block_ptr, (0, block_size))
-            grad_output_block_ptr = tl.advance(grad_output_block_ptr, (0, block_size))
+        for x_offset in range(0, x_size, x_block_size):
+            input = tl.load(input_block_ptr, boundary_check=(1,))
+            condition = tl.arange(0, x_block_size) + x_offset < x_size
+            input = tl.where(condition, input, -float("inf"))
+            peak = tl.where(condition, tl.maximum(max, input), 0)
+            max = peak
```
Comment (daemyung) [inline trident/kernel/softmax.py]: 계산된 다음에 업데이트가 되어야합니다.

File: trident/operation/attention.py (hunk 3 above)
Comment (daemyung) [reply trident/operation/attention.py:135]: 어텐션도 `stride`를 입력에서 받는 방향으로 업데이트 되어야 합니다.

File: trident/operation/softmax.py (hunk 1 above)
Comment (mejai1206) [reply trident/operation/softmax.py]: 음.. forward 도중에 발생하는 중간 결과물들도 해제하지 않나요?

File: trident/operation/attention.py (hunk 3 above)
Comment (mejai1206) [reply trident/operation/attention.py:135]: 넵 알겠습니다

File: trident/operation/softmax.py (hunk 1 above)
Comment (daemyung) [reply trident/operation/softmax.py]: 입력의 경우 기울기를 계산하기 위해 필요합니다.

**Review by mejai1206 — COMMENTED** [review]
LGTM

**Review by mejai1206 — APPROVED** [review]

File: trident/operation/softmax.py (hunk 1 above)
Comment (daemyung) [reply trident/operation/softmax.py]: 순전파시 `output`이 `input`에 의해 만들어져서 `input`은 저장하지 않고 역전파에서 `grad_input`을 `output`으로 생성하도록 변경했습니다.

**Review by kakao-steve-ai — APPROVED** [review]
리팩토링 & 디버깅 감사합니다
